geps/gep-1494/index.md

```diff
@@ -130,20 +130,383 @@ From @ongy, some additional goals to keep in mind:
+	//
+	// When this is set to HTTP, each backend must respond with a `200` status
+    // code in on a successful authorization. Any other code is considered
+	// an authorization failure.
+	//
+	// Feature Names:
+	// GRPC Support - HTTPRouteExtAuthGRPC
+	// HTTP Support - HTTPRouteExtAuthHTTP
+	//
+	// +unionDiscriminator
+	// +kubebuilder:validation:Enum=HTTP;GRPC
+	ExtAuthProtocol HTTPRouteExtAuthProtocol `json:"protocol"`
+
+	// BackendRefs is a list of references to backends to send authorization
+	// requests to.
+	//
+	// If this list contains more than one entry, authorization requests must
+	// be load-balanced across all the backends equally.
+	//
+	// The backends must speak the selected protocol (GRPC or HTTP) on the
+	// referenced port.
+	//
+	// If the backend service requires TLS, use BackendTLSPolicy to tell the
+	// implementation to supply the TLS details to be used to connect to that
+	// backend.
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=8
+	BackendRefs []BackendObjectReference `json:"backendRefs"`
+
+	// GRPCAuthConfig contains configuration for communication with ext_auth
+	// protocol-speaking backends.
+	//
+	// If unset, implementations must assume the default behavior for each
+	// included field is intended.
+	//
+	// +optional
+	GRPCAuthConfig *GRPCAuthConfig `json:"grpc,omitempty"`
+
+	// HTTPAuthConfig contains configuration for communication with HTTP-speaking
+	// backends.
+	//
+	// If unset, implementations must assume the default behavior for each
+	// included field is intended.
+	//
+	// +optional
+	HTTPAuthConfig *HTTPAuthConfig `json:"http,omitempty"`
+
+	// ForwardBody controls if requests to the authorization server should include
+	// the body of the client request; and if so, how big that body is allowed
+	// to be.
+	//
+	// It is expected that implementations will buffer the request body up to
+	// `forwardBody.maxSize` bytes. Bodies over that size must be rejected with a
+	// 4xx series error (413 or 403 are common examples), and fail processing
+	// of the filter.
+	//
+	// If unset, or `forwardBody.maxSize` is set to `0`, then the body will not
+	// be forwarded.
+	//
+	// Feature Name: HTTPRouteExtAuthForwardBody
+	//
+	// GEP Review Notes:
+	// Both Envoy and Traefik show support for this feature, but HAProxy and
+	// ingress-nginx do not. So this has a separate feature flag for it.
+	//
+	// +optional
+	ForwardBody *ForwardBodyConfig `json:"forwardBody,omitempty"`
+}
+
+// GRPCAuthConfig contains configuration for communication with ext_auth
+// protocol-speaking backends.
+type GRPCAuthConfig struct {
+
+	// AllowedRequestHeaders specifies what headers from the client request
+	// will be sent to the authorization server.
+	//
+	// If this list is empty, then all headers must be sent.
+	//
+	// +optional
+	// +kubebuilder:validation:MaxLength=64
+	AllowedRequestHeaders []string `json:"allowedHeaders,omitempty"`
```

Comment:
There are a lot of configuration options in ext_authz.proto. In the proposal, a few have been selected, e.g. alowed request headers (but not disallowed?), body settings, etc. but not other features (e.g. clearing route cache to influence service selection - very Envoy specific but also a very useful capability). What rubric is used to determine which capabilities to include and exclude?

Reply:
I've tried to keep it to "things I've definitely seen people ask for that also exist across multiple implementations". Part of the graduation requirement is for multiple implementations, preferably with multiple dataplanes, to implement this, so it's important that we try to keep it both small to start with and as portable as possible.

We can definitely add more stuff when people need it, this is me trying to pick a minimum viable feature set.